cosmwasm/enclaves/shared/contract-engine/src/contract_validation.rs

```diff
@@ -188,6 +201,75 @@ pub fn validate_contract_key(
     }
 }
 
+/// validate_contract_key validates the contract key against the contract address and code hash. If the contract was previously migrated, it also validates the contract key proof against the original contract key.
+pub fn validate_contract_key(
+    base_env: &BaseEnv,
+    canonical_contract_address: &CanonicalAddr,
+    contract_code: &ContractCode,
+) -> Result<(), EnclaveError> {
+    let og_contract_key: [u8; CONTRACT_KEY_LENGTH] = base_env.get_og_contract_key()?;
+
+    if base_env.was_migrated() {
+        trace!("Contract was migrated, validating proof");
+
+        let current_contract_key: [u8; CONTRACT_KEY_LENGTH] =
+            base_env.get_current_contract_key()?;
+
+        validate_current_contract_key(
+            &current_contract_key,
+            canonical_contract_address,
+            contract_code,
+            Some(&og_contract_key),
+        )?;
+
+        let sent_contract_key_proof = base_env.get_current_contract_key_proof()?;
+
+        let contract_key_proof = generate_contract_key_proof(
+            &canonical_contract_address.0 .0,
+            &contract_code.hash(),
+            &og_contract_key,
+            &current_contract_key, // this is already validated
+        );
+
+        if sent_contract_key_proof != contract_key_proof {
+            error!("Failed to validate contract key proof for a migrated contract");
+            return Err(EnclaveError::ValidationFailure);
+        }
+
+        Ok(())
+    } else {
+        trace!("Contract still has original code, validating contract_key");
+
+        validate_current_contract_key(
```

Comment:
Does this mean that a malicious actor can use an old version of a contract (which may be vulnerable to some bug), but still pass to it the new state of the upgraded contract? by specifying the old `og_contract_key`, `contract_code` and `address`, and setting `was_migrated` to `false`?

Reply:
> Does this mean that a malicious actor can use an old version of a contract (which may be vulnerable to some bug), but still pass to it the new state of the upgraded contract?

Yes, until we add state merkle proofs.